backend/danswer/db/document_set.py

```diff
@@ -0,0 +1,282 @@
+from sqlalchemy import delete
+from sqlalchemy import func
+from sqlalchemy import select
+from sqlalchemy.orm import Session
+
+from danswer.db.models import ConnectorCredentialPair
+from danswer.db.models import Document
+from danswer.db.models import DocumentByConnectorCredentialPair
+from danswer.db.models import DocumentSet as DocumentSetDBModel
+from danswer.db.models import DocumentSet__ConnectorCredentialPair
+from danswer.server.models import DocumentSetCreationRequest
+from danswer.server.models import DocumentSetUpdateRequest
+
+
+def _delete_document_set_cc_pairs(
+    db_session: Session, document_set_id: int, is_current: bool | None = None
+) -> None:
+    """NOTE: does not commit transaction, this must be done by the caller"""
+    stmt = delete(DocumentSet__ConnectorCredentialPair).where(
+        DocumentSet__ConnectorCredentialPair.document_set_id == document_set_id
+    )
+    if is_current is not None:
+        stmt = stmt.where(DocumentSet__ConnectorCredentialPair.is_current == is_current)
+    db_session.execute(stmt)
+
+
+def _mark_document_set_cc_pairs_as_outdated(
+    db_session: Session, document_set_id: int
+) -> None:
+    """NOTE: does not commit transaction, this must be done by the caller"""
+    stmt = select(DocumentSet__ConnectorCredentialPair).where(
+        DocumentSet__ConnectorCredentialPair.document_set_id == document_set_id
+    )
+    for row in db_session.scalars(stmt):
+        row.is_current = False
+
+
+def get_document_set_by_id(
+    db_session: Session, document_set_id: int
+) -> DocumentSetDBModel | None:
+    return db_session.scalar(
+        select(DocumentSetDBModel).where(DocumentSetDBModel.id == document_set_id)
+    )
+
+
+def insert_document_set(
+    document_set_creation_request: DocumentSetCreationRequest,
+    user_id: UUID | None,
+    db_session: Session,
+) -> tuple[DocumentSetDBModel, list[DocumentSet__ConnectorCredentialPair]]:
+    if not document_set_creation_request.cc_pair_ids:
+        raise ValueError("Cannot create a document set with no CC pairs")
+
+    # start a transaction
+    db_session.begin()
+
+    try:
+        new_document_set_row = DocumentSetDBModel(
+            name=document_set_creation_request.name,
+            description=document_set_creation_request.description,
+            user_id=user_id,
+        )
+        db_session.add(new_document_set_row)
+        db_session.flush()  # ensure the new document set gets assigned an ID
+
+        ds_cc_pairs = [
+            DocumentSet__ConnectorCredentialPair(
+                document_set_id=new_document_set_row.id,
+                connector_credential_pair_id=cc_pair_id,
+                is_current=True,
+            )
+            for cc_pair_id in document_set_creation_request.cc_pair_ids
+        ]
+        db_session.add_all(ds_cc_pairs)
+        db_session.commit()
+    except:
+        db_session.rollback()
+        raise
+
+    return new_document_set_row, ds_cc_pairs
+
+
+def update_document_set(
```

Comment:
Makes sense but a comment on the update document set + deleting expired doc set cc pair logic would be nice